jcl/src/java.base/share/classes/java/lang/invoke/MethodHandleResolver.java

```diff
@@ -360,6 +342,30 @@ private static final Object resolveInvokeDynamic(long j9class, String name, Stri
 /*[ENDIF] OPENJDK_METHODHANDLES*/
 	}
 
+	/**
+	 * Check access to the parameter and return classes within incoming MethodType.
+	 * 
+	 * @param lookup the Lookup object used for the access check.
+	 * @param type the incoming MethodType.
+	 * 
+	 * @throws IllegalAccessError if an inaccessible class is found.
+	 */
+	private static final void inaccessibleTypeCheck(Lookup lookup, MethodType type) throws IllegalAccessError {
+		int numElements = type.parameterCount() + 1;
+		Class<?>[] clsArray = new Class[numElements];
+		System.arraycopy(type.parameterArray(), 0, clsArray, 0, numElements - 1);
```

Comment:
This seems like an expensive way to iterate the MT parameterArray() as it requires multiple allocations:
* one in `type.parameterArray()`
* one in `new Class[numElements]`
And an arraycopy.

Is this only to be able to iterate thru using the `for each` syntax?

This could be done by using `type.parameterArray()` and iterate thru it followed by checking the return type.  Or even better, use the package private `MethodType#Class<?>[] ptypes()` method to avoid the allocation entirely.

```
Class<?> clsArray = type.ptypes();  // or type.parameterArray()
for (Class<?> cls : clsArray) {
	if (!cls.isPrimitive()) {
		if (!lookup.isClassAccessible(cls)) {
			/*[MSG "K0680", "Class '{0}' no access to: class '{1}'"]*/
			throw new IllegalAccessError(Msg.getString("K0680", lookup.lookupClass(), cls));
		}
	}
}
Class<?> cls = type.returnType();
if (!cls.isPrimitive()) {
if (!lookup.isClassAccessible(cls)) {
	/*[MSG "K0680", "Class '{0}' no access to: class '{1}'"]*/
	throw new IllegalAccessError(Msg.getString("K0680", lookup.lookupClass(), cls));
	}
}
```
Better to avoid the allocation even if duplicating the code for the check.

Reply:
Updated.